### PR TITLE
Template engine enhancement : context has microservices and iofogs list, and a filter to find agent of a microservice

### PR DESCRIPTION
**Pull request by franckOL**
- Does the iofog.org or README.md documentation need to change?
  - [ ] Yes 
  - [ ] No

### Description

 * Add to template engine context the list of microservices and lists of iofogs/agents
 * Add a template engine filter findAgent to be able to find agent of a microservice

TODO: need that microservices and iofogs lists retrieve in a more lazy way

### CODE REVIEW

**Comment by Pixcell** [comment]
looks good to me.
Could you add more tests to verify that the newly supported template values are being parsed correctly ?

**Comment by franckOL** [comment]
More unit test not easy, because need to mock controllers or services. I have a look but not so easy.
Integration tests check new features.

File: src/helpers/template-helper.js [hunk 1]
```diff
@@ -30,6 +43,32 @@ const rvaluesVarSubstition = async (subjects, templateContext) => {
...
+    let msvcEndpoint
+    let iofogListEndPoint
+    if (token) {
+      try {
+        msvcEndpoint = await MicroservicesController.getMicroservicesByApplicationEndPoint(req)
+        msvcEndpoint = msvcEndpoint.microservices
+        iofogListEndPoint = await FogController.getFogListEndPoint(req)
+        iofogListEndPoint = iofogListEndPoint.fogs
+      } catch (e) {
+        // Nothing to do, suppose the token has no permission to access. The is the case of agent
```
Comment (franckOL) [inline src/helpers/template-helper.js:83]: This part is due to post/put from agent. Their token throw an access error for getMicroservicesByApplicationEndPoint and getFogListEndPoint

==> need to do this middleware only for application and microservices post.

**Comment by Pixcell** [comment]
All good! Thank you.
Is this ready to merge? Or do you want to put more work into it?

**Comment by franckOL** [comment]
> All good! Thank you.
> Is this ready to merge? Or do you want to put more work into it?

What do you think about my comments is code ?
this is mergeable yet.

**Comment by Pixcell** [comment]
@franckOL 
We merged Application template integration into Controller. I had to refactor the templating logic to make it a proper middleware.
Could you update your PR and fix the conflicts? Thanks

**Comment by franckOL** [comment]
> @franckOL
> We merged Application template integration into Controller. I had to refactor the templating logic to make it a proper middleware.
> Could you update your PR and fix the conflicts? Thanks

OK done

**Comment by Pixcell** [comment]
@franckOL 
```
↳ Get Applications
  GET 127.0.0.1:51121/api/v3/application [200 OK, 2.78KB, 13ms]
  ✓  Status code is 200
  ✓  Response validation passed

↳ Get Application By Name
  GET 127.0.0.1:51121/api/v3/application/application-name [404 Not Found, 550B, 11ms]
  1. Status code is 200
  2. Response validation passed

↳ Update Application
  PATCH 127.0.0.1:51121/api/v3/application/application-name [404 Not Found, 536B, 63ms]
  3. Status code is 204

↳ Get Applications without system
  GET 127.0.0.1:51121/api/v3/application [200 OK, 2.78KB, 13ms]
  ✓  Status code is 200
  4. Response validation passed

↳ Delete Application
  DELETE 127.0.0.1:51121/api/v3/application/application-name-22 [204 No Content, 364B, 11ms]
  ✓  Status code is 204

↳ Delete Application With microservices
  DELETE 127.0.0.1:51121/api/v3/application/application-msvc-name [204 No Content, 364B, 65ms]
```

Postman tests are failing for `Application with template`
https://dev.azure.com/edgeworx/eclipse-iofog/_build/results?buildId=11024&view=logs&j=bcfce5c7-0bad-532d-898f-31dd6ba78dac&t=7c75fa6a-6578-55f9-a018-6c26addd6b43

From what I gather, it looks like it fails to create the application, since GET and PATCH return 404

Although the POSt to create returns success. I assume it is some postman variable naming issue

**Comment by franckOL** [comment]
> @franckOL
> 
> ```
> ↳ Get Applications
>   GET 127.0.0.1:51121/api/v3/application [200 OK, 2.78KB, 13ms]
>   ✓  Status code is 200
>   ✓  Response validation passed
> 
> ↳ Get Application By Name
>   GET 127.0.0.1:51121/api/v3/application/application-name [404 Not Found, 550B, 11ms]
>   1. Status code is 200
>   2. Response validation passed
> 
> ↳ Update Application
>   PATCH 127.0.0.1:51121/api/v3/application/application-name [404 Not Found, 536B, 63ms]
>   3. Status code is 204
> 
> ↳ Get Applications without system
>   GET 127.0.0.1:51121/api/v3/application [200 OK, 2.78KB, 13ms]
>   ✓  Status code is 200
>   4. Response validation passed
> 
> ↳ Delete Application
>   DELETE 127.0.0.1:51121/api/v3/application/application-name-22 [204 No Content, 364B, 11ms]
>   ✓  Status code is 204
> 
> ↳ Delete Application With microservices
>   DELETE 127.0.0.1:51121/api/v3/application/application-msvc-name [204 No Content, 364B, 65ms]
> ```
> 
> Postman tests are failing for `Application with template`
> https://dev.azure.com/edgeworx/eclipse-iofog/_build/results?buildId=11024&view=logs&j=bcfce5c7-0bad-532d-898f-31dd6ba78dac&t=7c75fa6a-6578-55f9-a018-6c26addd6b43
> 
> From what I gather, it looks like it fails to create the application, since GET and PATCH return 404
> 
> Although the POSt to create returns success. I assume it is some postman variable naming issue

I look a that, I see what is the problem.

**Pull request merged**